### PR TITLE
Fix 1013322

### PR DESCRIPTION
**Pull request by nosami**
Improve filtering for Model and DbContext class discovery && Make type discovery async

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1013322

### CODE REVIEW

**Comment by nosami** [comment]
@monojenkins backport release-8.4

**Comment by nosami** [comment]
@monojenkins rebase

**Comment by nosami** [comment]
@monojenkins rebase

File: main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderTemplateConfigurePage.cs [hunk 1]
```diff
@@ -69,21 +71,30 @@ protected override Widget GetMainControl ()
 						comboBox = new ComboBox ();
                     }
 
-					foreach (var option in comboField.Options) {
-						comboBox.Items.Add (option);
-					}
+					Task.Run (async () => {
```
[the remark this answers is not in the document]
Comment (rodrmoya) [reply main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderTemplateConfigurePage.cs]: Yeah, and the async API added should have `CancellationToken`'s passed to stop if the page is closed

[the remark this answers is not in the document]
Comment (nosami) [reply main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderTemplateConfigurePage.cs]: done

File: main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderArgs.cs [hunk 2]
```diff
@@ -43,5 +40,6 @@ public override int GetHashCode ()
 
 		public DotNetProject Project { get; internal set; }
 		public FilePath ParentFolder { get; internal set; }
+		public CancellationToken CancellationToken { get; internal set; }
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderArgs.cs]: where is this set?

Comment (nosami) [reply main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderArgs.cs]: here https://github.com/mono/monodevelop/pull/9208/files#diff-d526926b1924d24dc971601bd23cdf5bR42

File: main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderTemplateConfigurePage.cs [hunk 3]
```diff
@@ -65,38 +69,45 @@ protected override Widget GetMainControl ()
...
-                    }
-
-					foreach (var option in comboField.Options) {
-						comboBox.Items.Add (option);
 					}
 
+					Task.Run (async () => {
+						var options = await comboField.Options;
+						await Runtime.RunInMainThread (() =>
+						 Xwt.Toolkit.NativeEngine.Invoke (() => {
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderTemplateConfigurePage.cs]: do an:
```
if (cancellationToken.IsCancellationRequested) {
    return;
}
```

before the NativeEngine.Invoke call.

Comment (nosami) [reply main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Scaffolding/ScaffolderTemplateConfigurePage.cs]: added

**Comment by nosami** [comment]
@monojenkins backport release-8.4

**Pull request merged**